File: 2026/06-June/10-Wednesday/solution.py
import heapq
import math

class Solution:
    def maxTotalValue(self, nums: list[int], k: int) -> int:
        n = len(nums)
        if n == 0:
            return 0
        
        max_log = n.bit_length()
        max_st = [[0] * max_log for _ in range(n)]
        min_st = [[0] * max_log for _ in range(n)]
        
        for i in range(n):
            max_st[i][0] = nums[i]
            min_st[i][0] = nums[i]
            
        for j in range(1, max_log):
            for i in range(n - (1 << j) + 1):
                max_st[i][j] = max(max_st[i][j - 1], max_st[i + (1 << (j - 1))][j - 1])
                min_st[i][j] = min(min_st[i][j - 1], min_st[i + (1 << (j - 1))][j - 1])
                
        log_table = [0] * (n + 1)
        for i in range(2, n + 1):
            log_table[i] = log_table[i >> 1] + 1
            
        def get_range_value(l, r):
            length = r - l + 1
            j = log_table[length]
            mx = max(max_st[l][j], max_st[r - (1 << j) + 1][j])
            mn = min(min_st[l][j], min_st[r - (1 << j) + 1][j])
            return mx - mn
            
        pq = []
        for l in range(n):
            val = get_range_value(l, n - 1)
            # Use negative value because heapq is a min-heap
            heapq.heappush(pq, (-val, l, n - 1))
            
        total_sum = 0
        for _ in range(k):
            neg_val, l, r = heapq.heappop(pq)
            total_sum += (-neg_val)
            
            if r > l:
                next_r = r - 1
                next_val = get_range_value(l, next_r)
                heapq.heappush(pq, (-next_val, l, next_r))
                
        return total_sum

File: 2026/06-June/10-Wednesday/solution.java
import java.util.PriorityQueue;

class Solution {
    public long maxTotalValue(int[] nums, int k) {
        int n = nums.length;
        
        int[] log2 = new int[n + 1];
        for (int i = 2; i <= n; i++) {
            log2[i] = log2[i / 2] + 1;
        }
        
        int maxK = log2[n] + 1;
        int[][] maxST = new int[n][maxK];
        int[][] minST = new int[n][maxK];
        
        for (int i = 0; i < n; i++) {
            maxST[i][0] = nums[i];
            minST[i][0] = nums[i];
        }
        
        for (int j = 1; j < maxK; j++) {
            for (int i = 0; i + (1 << j) <= n; i++) {
                maxST[i][j] = Math.max(maxST[i][j - 1], maxST[i + (1 << (j - 1))][j - 1]);
                minST[i][j] = Math.min(minST[i][j - 1], minST[i + (1 << (j - 1))][j - 1]);
            }
        }
        
        PriorityQueue<long[]> pq = new PriorityQueue<>((a, b) -> Long.compare(b[0], a[0]));
        
        for (int l = 0; l < n; l++) {
            long val = getRangeValue(l, n - 1, maxST, minST, log2);
            pq.offer(new long[]{val, l, n - 1});
        }
        
        long totalSum = 0;
        
        for (int i = 0; i < k; i++) {
            long[] top = pq.poll();
            totalSum += top[0];
            
            int l = (int) top[1];
            int r = (int) top[2];
            
            if (r > l) {
                int nextR = r - 1;
                long nextVal = getRangeValue(l, nextR, maxST, minST, log2);
                pq.offer(new long[]{nextVal, l, nextR});
            }
        }
        
        return totalSum;
    }
    
    private long getRangeValue(int l, int r, int[][] maxST, int[][] minST, int[] log2) {
        int len = r - l + 1;
        int j = log2[len];
        
        int max = Math.max(maxST[l][j], maxST[r - (1 << j) + 1][j]);
        int min = Math.min(minST[l][j], minST[r - (1 << j) + 1][j]);
        
        return (long) max - min;
    }
}
